File: packages/elixir/lib/uncaught/prompt_builder.ex
defmodule Uncaught.PromptBuilder do
  @moduledoc """
  Fix-prompt builder — generates structured Markdown prompts for AI assistants.
  """

  @doc """
  Build a structured Markdown prompt for diagnosing and fixing the error.
  """
  def build(event) do
    sections = []

    # Intro
    sections =
      sections ++
        ["I have a production bug in my application that I need help diagnosing and fixing.\n"]

    # Error section
    sections =
      if event[:error] do
        error = event.error
        location = extract_location(error[:stack])
        lines = ["## Error", ""]
        lines = lines ++ ["- **Type:** #{error[:type] || "Error"}"]
        lines = lines ++ ["- **Message:** #{error[:message] || "(no message)"}"]

        lines =
          if location do
            lines ++ ["- **Location:** #{location}"]
          else
            lines
          end

        sections ++ [Enum.join(lines, "\n")]
      else
        sections
      end

    # Stack Trace
    stack_source = get_in_map(event, [:error, :resolvedStack]) || get_in_map(event, [:error, :stack])

    sections =
      if stack_source do
        frames =
          stack_source
          |> String.split("\n")
          |> Enum.take(15)
          |> Enum.map(&String.trim_trailing/1)
          |> Enum.join("\n")

        label =
          if get_in_map(event, [:error, :resolvedStack]) do
            "Stack Trace (source-mapped)"
          else
            "Stack Trace"
          end

        sections ++ ["## #{label}\n\n```\n#{frames}\n```"]
      else
        sections
      end

    # Failed Operation
    sections =
      if event[:operation] do
        sections ++ [format_operation(event.operation)]
      else
        sections
      end

    # HTTP Request Context
    sections =
      if event[:request] do
        sections ++ [format_request(event.request)]
      else
        sections
      end

    # User Session (last 5 breadcrumbs)
    sections =
      if event[:breadcrumbs] && length(event.breadcrumbs) > 0 do
        sections ++ [format_breadcrumbs(event.breadcrumbs)]
      else
        sections
      end

    # Environment
    sections =
      if event[:environment] do
        sections ++ [format_environment(event.environment)]
      else
        sections
      end

    # What I need
    sections =
      sections ++
        [
          Enum.join(
            [
              "## What I need",
              "",
              "1. **Root cause analysis** — explain why this error is occurring.",
              "2. **A fix** — provide the corrected code with an explanation of the changes.",
              "3. **Prevention** — suggest any guards or tests to prevent this from happening again."
            ],
            "\n"
          )
        ]

    Enum.join(sections, "\n\n") <> "\n"
  end

  # ---------------------------------------------------------------------------
  # Internal helpers
  # ---------------------------------------------------------------------------

  defp extract_location(nil), do: nil

  defp extract_location(stack) do
    stack
    |> String.split("\n")
    |> Enum.find_value(fn line ->
      trimmed = String.trim(line)

      # Elixir format: "(app) lib/file.ex:42: Module.function/2"
      case Regex.run(~r/^\((.+?)\)\s+(.+?:\d+)/, trimmed) do
        [_, _app, location] -> location
        _ ->
          # V8: "    at fn (file:line:col)"
          case Regex.run(~r/at\s+(?:.+?\s+\()?(.+?:\d+:\d+)\)?/, trimmed) do
            [_, location] -> location
            _ -> nil
          end
      end
    end)
  end

  defp format_operation(op) do
    lines = ["## Failed Operation", ""]
    lines = lines ++ ["- **Provider:** #{op[:provider] || ""}"]
    lines = lines ++ ["- **Type:** #{op[:type] || ""}"]
    lines = lines ++ ["- **Method:** #{op[:method] || ""}"]

    lines =
      if op[:params] do
        json = Jason.encode!(op.params, pretty: true)
        lines ++ ["- **Params:**", "```json", json, "```"]
      else
        lines
      end

    lines =
      if op[:errorCode] do
        lines ++ ["- **Error Code:** #{op.errorCode}"]
      else
        lines
      end

    lines =
      if op[:errorDetails] do
        lines ++ ["- **Error Details:** #{op.errorDetails}"]
      else
        lines
      end

    Enum.join(lines, "\n")
  end

  defp format_request(req) do
    lines = ["## HTTP Request Context", ""]

    lines =
      if req[:method] do
        lines ++ ["- **Method:** #{req.method}"]
      else
        lines
      end

    lines =
      if req[:url] do
        lines ++ ["- **URL:** #{req.url}"]
      else
        lines
      end

    lines =
      if req[:body] do
        body_str =
          if is_binary(req.body), do: req.body, else: Jason.encode!(req.body, pretty: true)

        lines ++ ["- **Body:**", "```json", body_str, "```"]
      else
        lines
      end

    Enum.join(lines, "\n")
  end

  defp format_breadcrumbs(crumbs) do
    recent = Enum.take(crumbs, -5)
    lines = ["## User Session", ""]

    lines =
      lines ++
        Enum.map(recent, fn crumb ->
          time = format_time(crumb[:timestamp] || "")
          type = crumb[:type] || "custom"
          message = crumb[:message] || ""
          "- `#{time}` **[#{type}]** #{message}"
        end)

    Enum.join(lines, "\n")
  end

  defp format_time(iso) do
    case Regex.run(~r/T(\d{2}:\d{2}:\d{2})/, iso) do
      [_, time] -> time
      _ -> iso
    end
  end

  defp format_environment(env) do
    lines = ["## Environment", ""]

    entries = [
      {"Deploy Environment", env[:deploy]},
      {"Framework", env[:framework]},
      {"Framework Version", env[:frameworkVersion]},
      {"Runtime", env[:runtime]},
      {"Runtime Version", env[:runtimeVersion]},
      {"Platform", env[:platform]},
      {"Browser",
       if(env[:browser],
         do: String.trim("#{env[:browser]} #{env[:browserVersion] || ""}"),
         else: nil
       )},
      {"OS", env[:os]},
      {"Device", env[:deviceType]},
      {"Locale", env[:locale]},
      {"Timezone", env[:timezone]},
      {"URL", env[:url]}
    ]

    lines =
      lines ++
        Enum.reduce(entries, [], fn {label, value}, acc ->
          if value && value != "" do
            acc ++ ["- **#{label}:** #{value}"]
          else
            acc
          end
        end)

    Enum.join(lines, "\n")
  end

  defp get_in_map(map, keys) when is_map(map) do
    Enum.reduce_while(keys, map, fn key, acc ->
      case acc do
        %{^key => value} -> {:cont, value}
        _ -> {:halt, nil}
      end
    end)
  end

  defp get_in_map(_, _), do: nil
end

File: packages/dotnet/src/Uncaught/PromptBuilder.cs
// ---------------------------------------------------------------------------
// Uncaught — fix-prompt builder
// ---------------------------------------------------------------------------

using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace Uncaught;

/// <summary>
/// Builds structured Markdown prompts for AI assistants to diagnose and fix errors.
/// </summary>
public static class PromptBuilder
{
    /// <summary>
    /// Build a structured Markdown prompt for the given event.
    /// </summary>
    public static string Build(UncaughtEvent ev)
    {
        var sections = new List<string>();

        // Intro
        sections.Add(
            "I have a production bug in my application that I need help diagnosing and fixing.\n");

        // Error section
        if (ev.Error != null)
        {
            var location = ExtractLocation(ev.Error.Stack);
            var lines = new List<string> { "## Error", "" };
            lines.Add($"- **Type:** {ev.Error.Type ?? "Error"}");
            lines.Add($"- **Message:** {ev.Error.Message ?? "(no message)"}");
            if (location != null)
            {
                lines.Add($"- **Location:** {location}");
            }
            sections.Add(string.Join("\n", lines));
        }

        // Stack Trace
        var stackSource = ev.Error?.ResolvedStack ?? ev.Error?.Stack;
        if (!string.IsNullOrEmpty(stackSource))
        {
            var frames = stackSource.Split('\n')
                .Take(15)
                .Select(l => l.TrimEnd());
            var label = ev.Error?.ResolvedStack != null
                ? "Stack Trace (source-mapped)"
                : "Stack Trace";
            sections.Add($"## {label}\n\n```\n{string.Join("\n", frames)}\n```");
        }

        // Failed Operation
        if (ev.Operation != null)
        {
            sections.Add(FormatOperation(ev.Operation));
        }

        // HTTP Request Context
        if (ev.Request != null)
        {
            sections.Add(FormatRequest(ev.Request));
        }

        // User Session (last 5 breadcrumbs)
        if (ev.Breadcrumbs?.Count > 0)
        {
            sections.Add(FormatBreadcrumbs(ev.Breadcrumbs));
        }

        // Environment
        if (ev.Environment != null)
        {
            sections.Add(FormatEnvironment(ev.Environment));
        }

        // What I need
        sections.Add(string.Join("\n", new[]
        {
            "## What I need",
            "",
            "1. **Root cause analysis** — explain why this error is occurring.",
            "2. **A fix** — provide the corrected code with an explanation of the changes.",
            "3. **Prevention** — suggest any guards or tests to prevent this from happening again."
        }));

        return string.Join("\n\n", sections) + "\n";
    }

    // -----------------------------------------------------------------------
    // Internal helpers
    // -----------------------------------------------------------------------

    private static readonly Regex ReDotnetLocation = new(
        @"^\s*at\s+.+?\s+in\s+(.+?:line\s+\d+)",
        RegexOptions.Compiled);

    private static readonly Regex ReV8Location = new(
        @"at\s+(?:.+?\s+\()?(.+?:\d+:\d+)\)?",
        RegexOptions.Compiled);

    private static string? ExtractLocation(string? stack)
    {
        if (string.IsNullOrEmpty(stack)) return null;

        foreach (var line in stack.Split('\n'))
        {
            var trimmed = line.Trim();

            var dotnetMatch = ReDotnetLocation.Match(trimmed);
            if (dotnetMatch.Success)
            {
                return dotnetMatch.Groups[1].Value;
            }

            var v8Match = ReV8Location.Match(trimmed);
            if (v8Match.Success)
            {
                return v8Match.Groups[1].Value;
            }
        }

        return null;
    }

    private static string FormatOperation(OperationInfo op)
    {
        var lines = new List<string> { "## Failed Operation", "" };
        lines.Add($"- **Provider:** {op.Provider}");
        lines.Add($"- **Type:** {op.OperationType}");
        lines.Add($"- **Method:** {op.Method}");
        if (op.Params != null)
        {
            lines.Add("- **Params:**");
            lines.Add("```json");
            lines.Add(JsonSerializer.Serialize(op.Params, new JsonSerializerOptions { WriteIndented = true }));
            lines.Add("```");
        }
        if (op.ErrorCode != null)
        {
            lines.Add($"- **Error Code:** {op.ErrorCode}");
        }
        if (op.ErrorDetails != null)
        {
            lines.Add($"- **Error Details:** {op.ErrorDetails}");
        }
        return string.Join("\n", lines);
    }

    private static string FormatRequest(RequestInfo req)
    {
        var lines = new List<string> { "## HTTP Request Context", "" };
        if (req.Method != null) lines.Add($"- **Method:** {req.Method}");
        if (req.Url != null) lines.Add($"- **URL:** {req.Url}");
        return string.Join("\n", lines);
    }

    private static string FormatBreadcrumbs(List<Breadcrumb> crumbs)
    {
        var recent = crumbs.TakeLast(5);
        var lines = new List<string> { "## User Session", "" };

        foreach (var crumb in recent)
        {
            var time = FormatTime(crumb.Timestamp);
            lines.Add($"- `{time}` **[{crumb.Type.ToString().ToLower()}]** {crumb.Message}");
        }

        return string.Join("\n", lines);
    }

    private static string FormatTime(string iso)
    {
        try
        {
            var dt = DateTimeOffset.Parse(iso);
            return dt.ToString("HH:mm:ss");
        }
        catch
        {
            return iso;
        }
    }

    private static string FormatEnvironment(EnvironmentInfo env)
    {
        var lines = new List<string> { "## Environment", "" };

        var entries = new (string Label, string? Value)[]
        {
            ("Deploy Environment", env.Deploy),
            ("Framework", env.Framework),
            ("Framework Version", env.FrameworkVersion),
            ("Runtime", env.Runtime),
            ("Runtime Version", env.RuntimeVersion),
            ("Platform", env.Platform),
            ("Browser", env.Browser != null
                ? $"{env.Browser} {env.BrowserVersion ?? ""}".Trim()
                : null),
            ("OS", env.Os),
            ("Device", env.DeviceType),
            ("Locale", env.Locale),
            ("Timezone", env.Timezone),
            ("URL", env.Url),
        };

        foreach (var (label, value) in entries)
        {
            if (!string.IsNullOrEmpty(value))
            {
                lines.Add($"- **{label}:** {value}");
            }
        }

        return string.Join("\n", lines);
    }
}
